packages/gatsby/src/bootstrap/requires-writer.ts

```diff
@@ -162,26 +162,41 @@ const getMatchPaths = (
 
 const createHash = (
   matchPaths: Array<IGatsbyPageMatchPath>,
-  components: Array<IGatsbyPageComponent>
+  components: Array<IGatsbyPageComponent>,
+  cleanedClientVisitedPageComponents: Array<IGatsbyPageComponent>
 ): string =>
   crypto
     .createHash(`md5`)
-    .update(JSON.stringify({ matchPaths, components }))
+    .update(
+      JSON.stringify({
+        matchPaths,
+        components,
+        cleanedClientVisitedPageComponents,
+      })
+    )
     .digest(`hex`)
 
 // Write out pages information.
 export const writeAll = async (state: IGatsbyState): Promise<boolean> => {
-  // console.log(`on requiresWriter progress`)
   const { program } = state
   const pages = [...state.pages.values()]
   const matchPaths = getMatchPaths(pages)
   const components = getComponents(pages)
+  const clientVisitedPageComponents = [...state.clientVisitedPages.values()]
+
+  // Remove any page components that no longer exist.
+  const cleanedClientVisitedPageComponents = components.filter(c =>
+    clientVisitedPageComponents.some(s => s === c.componentChunkName)
+  )
```

Comment:
let's go for better variable names
```suggestion
  const cleanedClientVisitedPageComponents = components.filter(component =>
    clientVisitedPageComponents.some(pageComponentChunkName => pageComponentChunkName === component.componentChunkName)
  )
```